### DISCUSSION TITLE
Return type typing impacted by type var name?

### DISCUSSION THREAD

**Discussion by Apakottur** [microsoft/pyright#7728] [Q&A] [answered]
I have the following code:
```python
# test.py
from typing import Any, Generic, TypeVar, overload


T = TypeVar("T")
S = TypeVar("S", bound=tuple[Any, ...])


class MyClass(Generic[T]):
    pass


@overload
async def inner(arg: MyClass[tuple[T]]) -> list[T]: ...


@overload
async def inner(arg: MyClass[S]) -> list[S]: ...


async def inner(arg: MyClass[tuple[T]] | MyClass[S]) -> list[T] | list[S]:
    return []


async def outer(arg: MyClass[tuple[T]] | MyClass[S]) -> list[T] | list[S]:
    return await inner(arg)

```

Running `pyright test.py` gives:
```text
test.py
  test.py:26:12 - error: Expression of type "list[tuple[T@outer]] | list[S@outer]" is incompatible with return type "list[T@outer] | list[S@outer]" (reportReturnType)
1 error, 0 warnings, 0 informations
```

I think that this error is incorrect, as the return type of `inner` is the same one as of `outer`.

I found two ways to make the error go away:
1. Remove all the async/await
2. Rename `S` to `T2`

So my questions:
1. Is my code typed incorrectly?
2. Is the async/await case a bug in pyright?
3. Is the type var name important/following some convention? I never imagined that renaming it would help.

**Comment by Apakottur** [comment]
Another way to make the error go away is to replace
```python
async def outer(arg: MyClass[tuple[T]] | MyClass[S]) -> list[T] | list[S]:
    return await inner(arg)
```

With
```python
async def outer(arg: MyClass[tuple[T]] | MyClass[S]) -> list[T] | list[S]:
    x = await inner(arg)
    return x
```

Which is logically the same.

**Reply by erictraut** [reply, marked as answer]
This has the same underlying cause as https://github.com/microsoft/pyright/issues/6651.

When pyright is performing bidirectional type inference for a function call, it pre-populates one or more TypeVar constraints based on the expected return type for the call. In cases where the return type is a union that includes multiple TypeVars, it can be ambiguous which constraints should be added. To break this ambiguity, pyright makes a best guess and proceeds with the evaluation with those constraints in place. To make operations deterministic, pyright internally sorts all of the subtypes within unions. It uses an internal "type comparison" algorithm that provides a deterministic order between types. This algorithm makes use of type names, which is why you're seeing the behavior change when you rename a TypeVar.

I don't think there's a need to open another issue on this topic since it's already covered in the issue I linked above. I don't currently have any good ideas how to fix this issue, but I'll continue to think about it. I'm able to reproduce similar issues in mypy (although its heuristics differ a bit, so the specific cases where they occur are different than in pyright).

**Reply by Apakottur** [reply]
Makes sense, very interesting!

For reference, some time ago I opened a somewhat similar issue on mypy - https://github.com/python/mypy/issues/15886.

Pyright does not raise an error on the mypy issue code.
